src/storage-controller/src/lib.rs

```diff
@@ -634,76 +580,60 @@ where
-
-            // Determine the intial since of the collection.
-            let initial_since = match storage_dependency {
-                Some(dep) => {
-                    let dependency_collection = self.collection(dep)?;
-                    let dependency_since = dependency_collection.implied_capability.clone();
-
-                    // If an item has a dependency, its initial since must be
-                    // advanced as far as its dependency, i.e. a dependency's
-                    // since may never be in advance of its dependents.
-                    //
-                    // We have to do this every time we initialize the
-                    // collection, though––the invariant might have been upheld
-                    // correctly in the previous epoch, but the
-                    // `data_shard_since` might not have compacted and, on
-                    // establishing a new persist connection, still have data we
-                    // said _could_ be compacted.
-                    if PartialOrder::less_than(&data_shard_since, &dependency_since) {
-                        // The dependency since cannot be in advance of the
-                        // dependent upper unless the collection is new. If the
-                        // dependency since advanced past the dependent's upper,
-                        // the dependent cannot read data from the dependency at
-                        // its upper.
-                        //
-                        // Another way of understanding that this is a problem
-                        // is that this means that the read hold installed on
-                        // the dependency was probably not been upheld––if it
-                        // were, the dependency's since could not have advanced
-                        // as far the dependent's upper.
-                        mz_ore::soft_assert_or_log!(
-                            write_frontier.elements() == &[T::minimum()]
-                                || PartialOrder::less_than(&dependency_since, write_frontier),
-                            "dependency ({dep}) since has advanced past dependent ({id}) upper \n
-                            dependent ({id}): since {:?}, upper {:?} \n
-                            dependency ({dep}): since {:?}",
-                            data_shard_since,
-                            write_frontier,
-                            dependency_since
-                        );
-
-                        dependency_since
-                    } else {
-                        data_shard_since
-                    }
+            let storage_dependencies =
+                self.determine_collection_dependencies(id, &description.data_source)?;
+
+            let dependency_read_holds = self
+                .storage_collections
+                .acquire_read_holds(storage_dependencies)
+                .expect("can acquire read holds");
```

Comment:
What happens when we are restarting envd and recreating collections in this `StorageCollections`, but before we acquire read holds for deps? Couldn't the compute controller downgrade the since through its `StorageCollections` if it boots before us?

Reply:
Currently that can't happen, because `create_collections` is always called from the `StorageController`. In the future, the protocol has to be that the `StorageController` get's a chance to run before anything else happens. And it (the `StorageController` that is "local" to that cluster/responsible for that cluster) will install read holds.

Also, for the future: there will be a `SinceHandle` per cluster, and on the cluster that is responsible for maintaining a collection, the `StorageController` acquires since holds through that. But the `SinceHandle` of other clusters doesn't need to be held back beyond what its `ComputeController`/the adapter have as requirements.